services/awsmeter/awsmeter.go

```diff
@@ -0,0 +1,146 @@
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/marketplacemetering"
+	"github.com/aws/aws-sdk-go/service/marketplacemetering/marketplacemeteringiface"
+)
+
+type AwsMeter struct {
+	store   store.Store
+	service *AWSMeterService
+	config  *model.Config
+}
+
+type AWSMeterService struct {
+	AwsDryRun      bool
+	AwsProductCode string
+	AwsMeteringSvc marketplacemeteringiface.MarketplaceMeteringAPI
+}
+
+func New(store store.Store, config *model.Config) *AwsMeter {
+	svc := &AWSMeterService{
+		AwsDryRun:      false,
+		AwsProductCode: "12345", //TODO
+	}
+
+	service, err := newAWSMarketplaceMeteringService()
+	if err != nil {
+		mlog.Error("newAWSMeterService", mlog.String("error", err.Error()))
+		return nil
+	}
+
+	svc.AwsMeteringSvc = service
+	return &AwsMeter{
+		store:   store,
+		service: svc,
+		config:  config,
+	}
+}
+
+func newAWSMarketplaceMeteringService() (*marketplacemetering.MarketplaceMetering, error) {
+	region := os.Getenv("AWS_REGION")
+	s := session.Must(session.NewSession(&aws.Config{Region: &region}))
+
+	creds := credentials.NewChainCredentials(
+		[]credentials.Provider{
+			&ec2rolecreds.EC2RoleProvider{
+				Client: ec2metadata.New(s),
+			},
+		})
+
+	_, err := creds.Get()
+	if err != nil {
+		mlog.Error("session is invalid", mlog.String("error", err.Error()))
+		return nil, errors.New("cannot obtain credentials")
+	}
+
+	return marketplacemetering.New(session.Must(session.NewSession(&aws.Config{
+		Credentials: creds,
+	}))), nil
+}
+
+// a report entry is for all metrics
+func (awsm *AwsMeter) GetUserCategoryUsage(dimensions []string, startTime time.Time, endTime time.Time) []*model.AWSMeterReport {
+	reports := make([]*model.AWSMeterReport, 0)
+
+	for _, dimension := range dimensions {
+		var userCount int64
+		var err error
+
+		switch dimension {
+		case model.AWS_METERING_DIMENSION_USAGE_HRS:
+			userCount, err = awsm.store.User().AnalyticsActiveCountForPeriod(model.GetMillisForTime(startTime), model.GetMillisForTime(endTime), model.UserCountOptions{})
+			mlog.Info("GetUserCategoryUsage", mlog.Int64("usercount", userCount), mlog.Err(err), mlog.Bool("bool", err != nil))
+			if err != nil {
+				mlog.Error("Failed to obtain usage data", mlog.String("dimension", dimension), mlog.String("start", startTime.String()), mlog.Int64("count", userCount), mlog.Err(err))
+			}
+		default:
+			mlog.Error("Dimension does not exist!", mlog.String("dimension", dimension))
+			err = errors.New("Dimension does not exist")
+		}
+
+		if err != nil {
+			mlog.Error("Failed to obtain usage.", mlog.String("dimension", dimension), mlog.Err(err))
+			return reports
+		}
+
+		report := &model.AWSMeterReport{
+			Dimension: dimension,
+			Value:     userCount,
+			Timestamp: startTime,
+		}
+
+		reports = append(reports, report)
+	}
+
+	return reports
+}
+
+func (awsm *AwsMeter) ReportUserCategoryUsage(reports []*model.AWSMeterReport) *model.AppError {
```

Comment:
Because this is a service, maybe makes way more sense to use plain errors instead of `AppError`, actually, in this case the error itself is not exposed to the user in any way, so can just use regular errors and everything should work out of the box.

Reply:
This applies to all the `AppError`s used in the whole service.

Reply:
sure, see comment above wrt to guidelines/etc. for this area of the codebase.